pkg/api/defaults.go

```diff
@@ -508,6 +512,101 @@ func (p *Properties) setHostedMasterProfileDefaults() {
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.EtcdDownloadURLBase != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.EtcdDownloadURLBase = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.EtcdDownloadURLBase
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubeBinariesSASURLBase != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.KubeBinariesSASURLBase = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubeBinariesSASURLBase
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubernetesImageBase != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.KubernetesImageBase = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubernetesImageBase
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.NVIDIAImageBase != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.NVIDIAImageBase = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.NVIDIAImageBase
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.TillerImageBase != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.TillerImageBase = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.TillerImageBase
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.VnetCNILinuxPluginsDownloadURL != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.VnetCNILinuxPluginsDownloadURL = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.VnetCNILinuxPluginsDownloadURL
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.VnetCNIWindowsPluginsDownloadURL != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.VnetCNIWindowsPluginsDownloadURL = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.VnetCNIWindowsPluginsDownloadURL
+			}
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.WindowsTelemetryGUID != "" {
+				azureStackCloudSpec.KubernetesSpecConfig.WindowsTelemetryGUID = p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.WindowsTelemetryGUID
+			}
+
+			//EndpointConfig
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.EndpointConfig.ResourceManagerVMDNSSuffix != "" {
+				azureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.AzureEnvironmentSpecConfig.EndpointConfig.ResourceManagerVMDNSSuffix
+			}
+			//OSImageConfig
+			azureStackCloudSpec.OSImageConfig = make(map[Distro]AzureOSImageConfig)
+			for k, v := range p.CustomCloudProfile.AzureEnvironmentSpecConfig.OSImageConfig {
+				azureStackCloudSpec.OSImageConfig[k] = v
+			}
+		} else {
+			p.CustomCloudProfile.AzureEnvironmentSpecConfig = &azureStackCloudSpec
+		}
+		AzureCloudSpecEnvMap[AzureStackCloud] = azureStackCloudSpec
```

Comment:
Thanks @honcao.

This is not ideal, but I think it works. @tariq1890 @CecileRobertMichon @mboersma thoughts here?

Basically, `AzureCloudSpecEnvMap[AzureStackCloud]` acts as a kind of global configuration for the Azure Stack context. Because we've never modified it, it kind of exists statically, in a const-like form (i.e., `AzureCloudSpecEnvMap`). Now that we're making that configuration surface area user-configurable, what's a practical iterative step forward?

Should we accept this as-is? Or should we wrap all this inside a light getter/setter interface?

Reply:
I think I'm missing something... what's the reason AzureCloudSpecEnvMap[AzureStackCloud] can't be defined in pkg/api/azenvtypes.go? defaults.go doesn't seem like the right place to do this at first glance

Reply:
It is defined in `pkg/api/azenvtypes.go`. This is a "set defaults setter" func for this particular config vector, mirroring other setters for other config vectors.

The question is what to do about the fact that `AzureCloudSpecEnvMap` is now, with this PR, becoming a "mutatable" configuration? Prior to this PR it acts as a "const"-like default hash table.

Reply:
So this points to how Azure Stack fundamentally differs as a Cloud environment when compared to the others. All the fields which are much immutable in normal Azure  Cloud Envs aren't necessarily invariant in Azure Stack environments. As each Azure  Stack environment is scoped within a customer's private cloud/data center, the fields will differ from customer to customer, unlike the Azure Public Cloud environments where fields are largely immutable and scoped within a broad domain.

Reply:
I will work on this in following PR